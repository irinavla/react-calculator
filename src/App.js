import React from 'react';
import './App.css';

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <p>
          this is the homepage
        </p>
      </header>
    </div>
  );
}

export default App;
