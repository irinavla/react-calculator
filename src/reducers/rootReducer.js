import * as actionTypes from '../actions/actionTypes';
import CalculatorOperations from './calculatorOperations';

const INITIAL_STATE = {
  value: null,
  displayValue: '0',
  operator: null,
  waitingForOperand: false
}

// helper function for updating the state
const updateObject = (oldObject, updadedProperties) => {
  return {
    ...oldObject,
    ...updadedProperties
  }
}

const inputDigit = (state, action) => {
  console.log('>>> INPUT DIGIT', state, action)

  const { displayValue, waitingForOperand } = state;
  let updatedState = null;

  if (waitingForOperand) {
      updatedState = {
          displayValue: String(action.digit),
          waitingForOperand: false
      }
      } else {
      updatedState = {
        displayValue: displayValue === '0' ? String(action.digit) : displayValue + action.digit
      }
  }
  return updateObject( state, updatedState );
}

const performOperation = (state, action) => {
  
  const { value, displayValue, operator } = state;
  const inputValue = parseFloat(displayValue);

  let updatedState = null;
  
  if (value == null) {
    updatedState = {
      value: inputValue,
      waitingForOperand: true,
      operator: action.operator
    }

    return updateObject(state, updatedState);

  } else if (operator) {
    const currentValue = value || 0;
    const newValue = CalculatorOperations[operator](currentValue, inputValue);

    updatedState = {
      value: newValue,
      displayValue: String(newValue),
      waitingForOperand: true,
      operator: action.operator
    }

    return updateObject(state, updatedState);
    
  }

  console.log('>>> PERFORM OPERATION', state, action)
};


const clearAll = () => {
  return INITIAL_STATE;
}


const rootReducer = (state = INITIAL_STATE, action) => {

  switch(action.type){
    case actionTypes.INPUT_DIGIT:
        return inputDigit(state, action);

    case actionTypes.PERFORM_OPERATION:
        return performOperation(state, action); 

    case actionTypes.CLEAR_ALL: 
        return clearAll();   

    default:
        return state;
}
}

export default rootReducer