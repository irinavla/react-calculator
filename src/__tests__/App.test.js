import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import App from '../App';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import rootReducer from '../reducers/rootReducer';

const store = createStore(
  rootReducer, 
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
  );

it('renders correctly', () => {
  const tree = renderer.create(<Provider store={store}> <App /></Provider>).toJSON();

  expect(tree).toMatchSnapshot();
});
