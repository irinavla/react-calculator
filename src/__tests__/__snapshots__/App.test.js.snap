// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  " ",
  <div
    className="App"
  >
    <div
      className="outer"
    >
      <div
        className="calculator"
      >
        <div
          className="result-display"
        >
          <div>
            0
          </div>
        </div>
        <div
          className="keypad"
        >
          <div
            className="keys"
          >
            <div
              className="clear vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              clear
            </div>
            <div
              className="numbers"
            >
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                7
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                8
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                9
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                4
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                5
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                6
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                1
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                2
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                3
              </div>
            </div>
          </div>
          <div
            className="operators"
          >
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              รท
            </div>
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              -
            </div>
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              +
            </div>
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              =
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`renders without crashing 1`] = `
Array [
  " ",
  <div
    className="App"
  >
    <div
      className="outer"
    >
      <div
        className="calculator"
      >
        <div
          className="result-display"
        >
          <div>
            0
          </div>
        </div>
        <div
          className="keypad"
        >
          <div
            className="keys"
          >
            <div
              className="clear vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              clear
            </div>
            <div
              className="numbers"
            >
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                7
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                8
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                9
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                4
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                5
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                6
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                1
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                2
              </div>
              <div
                className="key vh-center"
                onClick={[Function]}
                onTouchCancel={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                3
              </div>
            </div>
          </div>
          <div
            className="operators"
          >
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              รท
            </div>
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              -
            </div>
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              +
            </div>
            <div
              className="operator vh-center"
              onClick={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              =
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
